File: Sorting And Trees/Binary Search Trees, BST Sort/BST.cpp
// BST implementation - Insertion, Deletion, Searching, Traversals
// Author - rudrajit1729

#include<iostream>
using namespace std;

// Utility structure to represent Node of BST
struct Node{
    Node(int key): value(key), left(NULL), right(NULL){}
    int value;
    Node *left;
    Node *right;
};

// BST class
class BST{
    private:
        Node *root;
        Node* insert(Node *node, int val);
        Node* deleteNode(Node *node, int val);
        int findNode(Node *node, int key);
        void inorder(Node *node);
        void preorder(Node *node);
        void postorder(Node *node);
    public:
        BST(); // Constructor
        //~BST(); // Destructor
        Node* minValNode(Node *node);
        Node* insert(int val){root = insert(root, val);}
        Node* deleteNode(int val){root = deleteNode(root, val);}
        void inorder(){inorder(root);}
        void preorder(){preorder(root);}
        void postorder(){postorder(root);}
        int findNode(int key){findNode(root, key);}
};

BST::BST()
{
    root = NULL;
}

// Utility function to add nodes in BST
Node* BST::insert(Node *root, int val)
{
    // Base case
    if (root == NULL){
        root = new Node(val);
        return root;
        
    }
    // If value less than root, it goes in left subtree
    if (root->value > val)
    {
        if (root->left == NULL)
            root->left = new Node(val);
        else
            root->left = insert(root->left, val);
    }
    // Else goes in right subtree
    else
    {
        if (root->right == NULL)
            root->right = new Node(val);
        else
            root->right = insert(root->right, val);
    }
    return root;
}

// Utility functions for traversals
void BST::inorder(Node *root)
{
    if (root != NULL)
    {
        inorder(root->left);
        cout<<root->value<<"\t";
        inorder(root->right);
    }
}

void BST::preorder(Node *root)
{
    if (root != NULL)
    {
        cout<<root->value<<"\t";
        preorder(root->left);
        preorder(root->right);
    }
}

void BST::postorder(Node *root)
{
    if (root != NULL)
    {
        postorder(root->left);
        postorder(root->right);
        cout<<root->value<<"\t";
    }
}

// Utility function to find presence of node
int BST::findNode(Node *root, int key)
{
    // Base case
    if (root==NULL) return 0;
    // If key less than root value look in left subtree
    if (root->value < key)
        return findNode(root->left, key);
    // If key greater than root value look in right subtree
    else if(root->value > key)
        return findNode(root->right, key);
    // Else found
    else
        return 1;
}

// Getting minimum Value node w.r.t. a node
Node *BST:: minValNode(Node *root)
{
    Node *curr = root;
    while(curr->left != NULL)
        curr = curr->left;
    return curr;
}
// Utility function to delete a node from BST
Node*BST:: deleteNode(Node *root, int key)
{
    // Base Case
    if (root == NULL) return root;

    // If key is smaller than root value look in left subtree
    if (root->value > key)
        root->left = deleteNode(root->left, key);
    // Else if key is larger than root value look in right subtree
    else if (root->value < key)
        root->right = deleteNode(root->right, key);
    // Else present node is the node to be deleted
    else
    {
        // If node has one or no children
        if (root->left == NULL)
        {
            Node * temp = root->right;
            root = NULL;
            return temp;
        }
        else if (root->right == NULL)
        {
            Node * temp = root->left;
            root = NULL;
            return temp;
        }
        // Node has two children
        // Replace root by inorder successor - Lowest key in right subtree
        else
        {
            // Get inorder successor
            Node * temp = minValNode(root->right);
            
            // Copy value from temp into root
            root->value = temp->value;
            // Delete temp value from right subtree
            root->right = deleteNode(root->right, temp->value);
        }
        return root;
    }
    
}


int main()
{
    /*

                  49
               /      \
             46       79
            /  \      /  \
         43   45     64  83

    */
    BST tree;
    // Insertion
    tree.insert(49);
    tree.insert(46);
    tree.insert(43);
    tree.insert(45);
    tree.insert(79);
    tree.insert(64);
    tree.insert(83);
    
    //Traversals
    cout<<"\nInorder\n";
    tree.inorder();
    cout<<"\nPreorder\n";
    tree.preorder();
    cout<<"\nPostorder\n";
    tree.postorder();   

    // Searching
    cout<<"\nFinding 46 : "<<tree.findNode(46)<<endl;
    cout<<"Finding 78 : "<<tree.findNode(78)<<endl;

    // Deleting node
    tree.deleteNode(43);//No child
    cout<<"Inorder:\n";
    tree.inorder();
    tree.deleteNode(46); // 1 child
    cout<<"\nInorder:\n";
    tree.inorder();
    tree.deleteNode(79); // 2 children
    cout<<"\nInorder:\n";
    tree.inorder();

}

File: Sorting And Trees/Binary Search Trees, BST Sort/BST.py
# BST implementation - insertion, finding, traversals, deletion
# Author - rudrajit1729

# Utility class represents individual nodes in BST
class Node:
    def __init__(self, key = 0):
        self.left = None
        self.right = None
        self.value = key 

# Utility function to insert node in BST
def insert(root, value):
    # Base Case
    if root is None:
        return Node(value)
    # Value smaller than root, goes into left subtree
    if root.value > value:
        if root.left is None:
            root.left = Node(value)
        else:
            root.left = insert(root.left, value)
    # value larger than or equal to root, goes into right subtree
    else:
        if root.right is None:
            root.right = Node(value)
        else:
            root.right = insert(root.right, value)
    return root

# Utility functions for traversals
def inorder(root):
    if root:
        inorder(root.left)
        print(root.value, end = " ")
        inorder(root.right)

def preorder(root):
    if root:
        print(root.value, end = " ")
        preorder(root.left)
        preorder(root.right)

def postorder(root):
    if root:
        postorder(root.left)
        postorder(root.right)
        print(root.value, end = " ")

# Getting the minimum value node of the tree
def minValNode(root):
    curr = root
    while curr.left:
        curr = curr.left
    return curr

# Utility function for deleting a node
def delete(root, key):
    # Base Case
    if root is None:
        return root
    # If key is smaller than root value it's in left subtree
    if key < root.value:
        root.left = delete(root.left, key)
    # If key is larger than root value it's in right subtree
    elif key > root.value:
        root.right = delete(root.right, key)
    # Key matches the root value
    else:
        # Node with one or no child
        if root.left is None:
            temp = root.right
            root = None
            return temp
        elif root.right is None:
            temp = root.left
            root = None
            return temp
        
        # Node with two children: Replace by inorder successor(smallest in right subtree)
        temp = minValNode(root.right)
        # Value of the inorder successor stored in node
        root.value = temp.value
        # Delete the inorder successor from the right subtree
        root.right = delete(root.right, temp.value)

    return root

# Utility function for finding a key in BST
def findNode(root, key):
    if root is None:
        return False
    if key < root.value:
        return findNode(root.left, key)
    elif key > root.value:
        return findNode(root.right, key)
    else:
        return True

# Demonstration method
def main():
    r'''

                  49
               /      \
             46       79
            /  \      /  \
         43   45     64  83

    '''
    root = None
    root = insert(root, 49)
    root = insert(root, 46)
    root = insert(root, 43)
    root = insert(root, 45)
    root = insert(root, 79)
    root = insert(root, 64)
    root = insert(root, 83)
    print("In order traversal")
    inorder(root)
    root = delete(root, 43)
    print("\nIn order traversal")
    inorder(root)
    root = delete(root, 46)
    print("\nIn order traversal")
    inorder(root)
    root = delete(root, 79)
    print("\nIn order traversal")
    inorder(root)

    print("\nFinding a node demo:\n")
    print("Finding 83 : ", findNode(root, 83))
    print("\nFinding 43 : ", findNode(root, 43))


if __name__ == "__main__":
    main()